File: khsexk-Java/src/BOJ/BOJ_2531.java
package BOJ;

import java.io.*;

public class BOJ_2531 {
	static int max = Integer.MIN_VALUE;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] s = br.readLine().split(" ");
		int N = Integer.parseInt(s[0]);
		int d = Integer.parseInt(s[1]);
		int k = Integer.parseInt(s[2]);
		int c = Integer.parseInt(s[3]);
		
		int[] sushi = new int[N];
		
		
		for(int i=0 ; i<N ; i++)
			sushi[i] = Integer.parseInt(br.readLine());
		
		for(int i=0 ; i<N ; i++) {
			boolean[] eatedSushi = new boolean[d+1];
			int cnt = 0;
			int idx = i;
			int sum = 0;
			
			while(cnt++ < k) {
				if(!eatedSushi[sushi[idx]]) sum++;
				eatedSushi[sushi[idx]] = true;
				if(idx == N-1) idx = 0;
				else idx++;
			}
			if(!eatedSushi[c]) sum ++;
			
			max = (sum > max) ? sum:max;
		} // for_i
		
		System.out.println(max);
	}

}

File: yongshub-python/Two Pointer/BOJ_2531.py
import sys
input = sys.stdin.readline
# 접시 수, 가짓수, 연속해서 먹는 수, 쿠본 번호
N, d, k, c = map(int, input().split())
sushi = []
# 벨트 초기화
for _ in range(N):
    sushi.append(int(input().rstrip()))

# 연속해서 먹는 접시의 수 k 경우의 수 구하기
result = 0
for i in range(N):
    end = i  # 처음 가리키는 end 인덱스
    cnt = 0
    case = set()
    while cnt < k:  # cnt가 k보다 작을 때
        case.add(sushi[end % N])
        end += 1
        cnt += 1
    case.add(c)
    result = max(result, len(case))
print(result)
# 원형 큐로 생각해야한다!
